libyul/YulName.cpp

```diff
@@ -0,0 +1,475 @@
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libyul/optimiser/NameCollector.h>
+
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/AST.h>
+
+#include <fmt/compile.h>
+
+namespace solidity::yul
+{
+
+struct YulNameRepository::IndexBoundaries
+{
+	size_t beginTypes {};
+	size_t endTypes {};
+	size_t beginBuiltins {};
+	size_t endBuiltins {};
+};
+
+YulNameRepository::YulNameRepository(solidity::yul::Dialect const& _dialect):
+	m_dialect(_dialect), m_indexBoundaries(std::make_unique<IndexBoundaries>())
+{
+	defineName("");
+
+	for (auto const& type: _dialect.types)
+		if (type.empty())
+		{
+			m_indexBoundaries->beginTypes = 0;
+			m_dialectTypes.emplace_back(emptyName(), type.str());
+		}
+		else
+		{
+			m_indexBoundaries->beginTypes = 1;
+			m_dialectTypes.emplace_back(defineName(type.str()), type.str());
+		}
+	m_indexBoundaries->endTypes = m_index;
+	m_indexBoundaries->beginBuiltins = m_index;
+
+	auto const& builtinNames = _dialect.builtinNames();
+	m_predefined.verbatim = defineName("@ verbatim");
+	for (auto const& label: builtinNames)
+		if (!label.empty())
+		{
+			auto const name = defineName(label);
+			if (auto const* function = m_dialect.builtin(YulString(label)))
+				m_builtinFunctions[name] = convertBuiltinFunction(name, *function);
+		}
+	m_indexBoundaries->endBuiltins = m_index;
+
+	m_predefined.boolType = nameOfType(_dialect.boolType.str());
+	m_predefined.defaultType = nameOfType(_dialect.defaultType.str());
+	if (builtinNames.count("dataoffset") > 0)
+		m_predefined.dataoffset = nameOfBuiltin("dataoffset");
+	else
+		m_predefined.dataoffset = defineName("dataoffset");
+	if (builtinNames.count("datasize") > 0)
+		m_predefined.datasize = nameOfBuiltin("datasize");
+	else
+		m_predefined.datasize = defineName("datasize");
+	if (builtinNames.count("selfdestruct") > 0)
+		m_predefined.selfdestruct = nameOfBuiltin("selfdestruct");
+	else
+		m_predefined.selfdestruct = defineName("selfdestruct");
+	if (builtinNames.count("tstore") > 0)
+		m_predefined.tstore = nameOfBuiltin("tstore");
+	else
+		m_predefined.tstore = defineName("tstore");
+	if (builtinNames.count("memoryguard") > 0)
+		m_predefined.memoryguard = nameOfBuiltin("memoryguard");
+	else
+		m_predefined.memoryguard = defineName("memoryguard");
+	if (builtinNames.count("eq") > 0)
+		m_predefined.eq = nameOfBuiltin("eq");
+	else
+		m_predefined.eq = defineName("eq");
+	if (builtinNames.count("add") > 0)
+		m_predefined.add = nameOfBuiltin("add");
+	else
+		m_predefined.add = defineName("add");
+	if (builtinNames.count("sub") > 0)
+		m_predefined.sub = nameOfBuiltin("sub");
+	else
+		m_predefined.sub = defineName("sub");
```

Comment:
Could this be put into a function? It would improve readability.

Maybe something like:
```c++
	auto setPredefined = [&](std::string const _label, YulName& _name)
	{
		if (builtinNames.count(_label) > 0)
			_name = nameOfBuiltin(_label);
		else
			_name = defineName(_label);
	};

	setPredefined("dataoffset", m_predefined.dataoffset);
	setPredefined("datasize", m_predefined.datasize);
	setPredefined("selfdestruct", m_predefined.selfdestruct);
	setPredefined("tstore", m_predefined.tstore);
	setPredefined("memoryguard", m_predefined.memoryguard);
	setPredefined("eq", m_predefined.eq);
	setPredefined("add", m_predefined.add);
	setPredefined("sub", m_predefined.sub);
```

Reply:
true, much more readable that way:)